src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -107,6 +107,23 @@ public override Type ExpectedClrType
             get { return typeof(TStructuralType); }
         }
 
+        /// <summary>
+        /// The list of property names that can be updated.
+        /// </summary>
+        /// <remarks>When the list is modified, any modified properties that were removed from the list are no longer
+        /// considered to be changed.</remarks>
+        public IEnumerable<string> UpdatableProperties
```

Comment:
I'd prefer to have an overload of "Patch of Put " to accept the list of updatable properties.
Because we have the overload of constructor to accept the list of updatable properties. This properties doesn't make sense here.

@mikepizzo What do you think?

Reply:
@xuzhg @dxrdxr shared reasons why they believe an overload would not be suitable in the issue discussion here: https://github.com/OData/WebApi/issues/2478#issuecomment-828337073

Reply:
@xuzhg In addition to my comments in #2478, the existing `updatableProperties` is set only through the constructor, a very strange pattern. It is a write-only state of the class that cannot be examined through any interface on the class. Operations such as `TrySetPropertyValue()` can fail but there is no way to determine if it failed because it was not updatable, or another reason (e.g. type mismatch), this is a developer headache (note that the current docs do not specify that it will fail if not updatable).

Additionally, the current ctor argument only initialization goes against the much more common practice of allowing state (properties) to be set either as an argument or as with an object initializer.
```
var delta = new Delta<Student>(typeof(Student), myUpdatableProperties);
```
```
var delta = new Delta<Student>()
{
    UpdatableProperties = myUpdatableProperties
};
```
Lastly as I said in my other comments, #3, for consistent semantics it is not just `Put` and `Patch` it impacts `GetChangedPropertyNames` and `GetUnchangedPropertyNames`. Having the developer maintain and wrap all of the Get/Set methods to filter properties versus pushing it close to the object that is already managing the state, is a cleaner approach.

Reply:
@xuzhg could you provide your views on this one?

Reply:
I think it's ok to expose the set for property. Or maybe we can add a new method to update the "UpdatableProperties".

Reply:
@xuzhg 
> Or maybe we can add a new method to update the "UpdatableProperties".

To create a new method to update the field it would require a `GetUpdateableProperties` and a `SetUpdateableProperties`. How is that different that the getter/setter on the property `UpdateableProperties`?

It is important to fix the problem that the existing `_updatableProperties` has no way retrieve its current value, a write-only field. So a Set method or property setter must be part of the solution.

Reply:
I think keeping it as a property makes more sense.

Reply:
> I'd prefer to have an overload of "Patch of Put " to accept the list of updatable properties.
> Because we have the overload of constructor to accept the list of updatable properties. This properties doesn't make sense here.
> 
> @mikepizzo What do you think?

@xuzhg Do you still need changes to be made?

Reply:
Excellent question! 

I was following the existing implementation being an `IEnumerable` so left it as a set the whole property.

It makes mores sense to expose it as an `IList` so that it can be updated instead of just replaced.

I will reimplement this and see the impact.

Thanks!

Reply:
Making it `IList` and readonly does make a lot of sense

Reply:
I updated the Pull Request to change to having it be an `IList` and readonly.

This change invalidates an existing assumption that the contents of the `_updatableProperties` field only contains the names of properties that are also in the `allProperties` dictionary. As an `IList`, any arbitrary string can be added, so the code needed to be updated in places to filter against invalid properties.

Additionally, the internal use of `HashSet` must be changed to `List` as `HashSet` does not implement `IList`. This will have a minor performance impact on the use of `Contains` going from O(1) to O(n), however the number of properties is always small so should not be a factor.